### PR TITLE
[Flux] Enable classifier-free guidance for training and inference

### PR DESCRIPTION
**Pull request by wwwjn**
## Context
1. Refactor dataloader to support classifier-free guidance: Replace some of the strings with empty strings.
2. Refactor test_generate_image.py to run a forward pass to generate image with classifer-free guidance.
```
# sampling pseudocode given prompt (e.g. "a picture of a cat") :
1. sample noise of shape [B, 64, height  // 16, width // 16]
2. duplicate noise i.e. x = torch.cat((noise, noise))
3. patchify x
4. choose timesteps, e.g., t = torch.linspace(1.0, 0.0, 30)
5. compute t5 / clip embeddings for prompt and empty string "" (you could also cache the empty string embeddings if you want )
6. for loop over timesteps for t_curr, t_prev in zip(timesteps[:-1], timesteps[1:]):
    evaluate pred = model(x, t, t5_embeddings, clip_embeddings)
    classifier-free guidance pred_u, pred_c = pred.chunk(2) -> pred = pred_u + cfg_val * (pred_c - pred_u)
    advanced x =  x  + (t_prev - t_curr) * pred
7. image = autoencoder(x)
```
3. Remove guided-distillation related params.
4. Refactor inference related code to `sampling.py`. Adding periodical eval in `train.py`
5. Seed management for Flux model
6. Adding basic eval function (generate image every few steps) during training

##  Seed management during Flux training and eval

With in 1 single training run: 
1. During training, each `dp_shard` rank should have different random seed. Because we are randomly dropout the prompt information to train a unconditional + conditional model. We don't want to each rank drop samples from same position in dataloader.
2. During evaluation/sampling, since we are generating from random noise, so we want to use the exact same seed for each evaluation step.
    - Run image sampling on Rank 0 only, Use the local random seed on rank0 to generate

Between 2 different training runs: 
- We want to reproduce the seed for the first time to strictly reproduce the 1st run results.

## Next Step:
- Adding more evaluation functionality: Calculate loss fix `t` value .

### CODE REVIEW

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 1]
```diff
@@ -10,12 +10,11 @@
 from typing import Any, Callable, Optional
 
 import numpy as np
+import PIL.Image
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]:
```suggestion
import PIL
```
since you only use PIL directly?

File: torchtitan/experiments/flux/train_configs/flux_dev_model.toml [hunk 2]
```diff
@@ -37,15 +37,25 @@ seq_len = 512
...
+seed = 0
+classifer_free_guidance_prob = 0.1
 
 [encoder]
 t5_encoder = "google/t5-v1_1-xxl"
 clip_encoder = "openai/clip-vit-large-patch14"
 max_t5_encoding_len = 4096
 auto_encoder_path = "torchtitan/experiments/flux/assets/autoencoder/ae.safetensors"  # Autoencoder to use for image
 
+[sampling]
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train_configs/flux_dev_model.toml]: maybe rename to eval, since sampling itself does not necessarily relate to training

File: torchtitan/experiments/flux/train_configs/flux_dev_model.toml [hunk 3]
```diff
@@ -37,15 +37,25 @@ seq_len = 512
 max_norm = 1.0  # grad norm clipping
 steps = 300_000
 compile = false
-dataset = "cc12m"
-guidance = 3.5
+dataset = "cc12m-wds"
+seed = 0
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train_configs/flux_dev_model.toml]: let's not fix seed in toml files -- they probably should be used for debugging only

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 4]
```diff
@@ -205,6 +206,16 @@ def __iter__(self):
                     )
                     continue
 
+                # Classifier-free guidance: Replace some of the strings with empty strings.
+                # Using random seed to ensure each rank has the same dropouted samples when TP and CP are enabled.
+                seed = self.job_config.training.seed
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: Please be careful when managing seed usage. Let's do a global audit and have a clear story.

File: torchtitan/distributed/utils.py [hunk 5]
```diff
@@ -46,6 +46,7 @@ def set_determinism(
     device: torch.device,
     seed: int | None = None,
     deterministic: bool = False,
+    distinct_seed_mesh_dims: str = "pp",
 ) -> None:
     """
     Set the same DTensor manual seed for all ranks within the same DTensor SPMD group, but different
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py]: pls also update docstring to be more general, and still include SPMD + PP as the archetypal example

File: torchtitan/distributed/utils.py [hunk 6]
```diff
@@ -46,6 +46,7 @@ def set_determinism(
     device: torch.device,
     seed: int | None = None,
     deterministic: bool = False,
+    distinct_seed_mesh_dims: str = "pp",
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py]: maybe
```suggestion
    distinct_seed_mesh_dim: str = "pp",
```

File: torchtitan/distributed/utils.py [hunk 7]
```diff
@@ -81,22 +82,31 @@ def set_determinism(
...
         logger.debug(
-            f"PP rank {pp_mesh.get_local_rank()}, Global rank {c10d.get_rank()} using seed: {seed}"
+            f"{distinct_seed_mesh_dims} rank {distinct_mesh.get_local_rank()}, Global rank {c10d.get_rank()} using seed: {seed}"
         )
-        spmd_mesh_dims = list(
+        duplicate_seed_mesh_dims = list(
             filter(lambda name: name != "pp", world_mesh.mesh_dim_names)
         )
-        spmd_mesh = world_mesh[spmd_mesh_dims] if len(spmd_mesh_dims) else None
+        duplicate_seed_mesh_dims = (
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py]: rename to
```suggestion
        duplicate_seed_mesh = (
```

File: torchtitan/distributed/utils.py [hunk 8]
```diff
@@ -81,22 +82,31 @@ def set_determinism(
...
+        seed += distinct_mesh.get_local_rank()
         seed %= 2**64
 
         logger.debug(
-            f"PP rank {pp_mesh.get_local_rank()}, Global rank {c10d.get_rank()} using seed: {seed}"
+            f"{distinct_seed_mesh_dims} rank {distinct_mesh.get_local_rank()}, Global rank {c10d.get_rank()} using seed: {seed}"
         )
-        spmd_mesh_dims = list(
+        duplicate_seed_mesh_dims = list(
             filter(lambda name: name != "pp", world_mesh.mesh_dim_names)
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py]:
```suggestion
            filter(lambda name: name != distinct_seed_mesh_dim, world_mesh.mesh_dim_names)
```

File: torchtitan/experiments/flux/flux_argparser.py [hunk 9]
```diff
@@ -45,3 +45,45 @@ def extend_parser(parser: argparse.ArgumentParser) -> None:
...
+    parser.add_argument(
+        "--eval.classifier_free_guidance_scale",
+        type=float,
+        default=5.0,
+        help="Classifier-free guidance scale when sampling",
+    )
+    parser.add_argument(
+        "--eval.sample_img_width",
+        type=int,
+        default=512,
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/flux_argparser.py]: I'd suggest setting eval default no larger than the train default (256?).

File: torchtitan/experiments/flux/flux_argparser.py [hunk 10]
```diff
@@ -45,3 +45,45 @@ def extend_parser(parser: argparse.ArgumentParser) -> None:
...
+        help="Image height to sample",
+    )
+    parser.add_argument(
+        "--eval.denoising_steps",
+        type=int,
+        default=512,
+        help="How many denoising steps to sample when generating an image",
+    )
+    parser.add_argument(
+        "--eval.output_dir",
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/flux_argparser.py]: could you follow torchtitan convention and use it in concatenation of the `job.dump_dir`?
example: https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/metrics.py#L254-L257

File: torchtitan/experiments/flux/tests/test_generate_image.py [hunk 11]
```diff
@@ -80,177 +34,109 @@ def test_generate_image(self):
...
+                "0",
+                "--training.classifer_free_guidance_prob",
+                "0.1",
+                "--encoder.t5_encoder",
+                "google/t5-v1_1-base",
+                "--encoder.clip_encoder",
+                "openai/clip-vit-large-patch14",
+                "--encoder.max_t5_encoding_len",
+                "512",
+                # sampling params
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/test_generate_image.py]: "sampling" vs. "eval", let's make them consistent

File: torchtitan/experiments/flux/train.py [hunk 12]
```diff
@@ -155,19 +161,18 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
             timesteps = torch.rand((bsz,)).to(labels)
             sigmas = timesteps.view(-1, 1, 1, 1)
             noisy_latents = (1 - sigmas) * labels + sigmas * noise
-            guidance = torch.full((bsz,), self._guidance).to(labels)
 
         target = noise - labels
 
         assert len(model_parts) == 1
 
+        model = self.model_parts[0].train().requires_grad_(True)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: why do we need to explicitly set `requires_grad_(True)`?

File: torchtitan/experiments/flux/train_configs/flux_schnell_model.toml [hunk 13]
```diff
@@ -38,15 +37,24 @@ seq_len = 512
...
 clip_encoder = "openai/clip-vit-large-patch14"
 max_t5_encoding_len = 4096
 auto_encoder_path = "torchtitan/experiments/flux/assets/autoencoder/ae.safetensors"  # Autoencoder to use for image
 
+[eval]
+enable_classifer_free_guidance = true
+classifer_free_guidance_scale = 5.0
+sample_img_width = 512
+sample_img_height = 512
+denoising_steps = 30
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train_configs/flux_schnell_model.toml]: What is the suggested default? If it is always 30, we probably should change to 30, and maybe remove it from toml if it is not expected to be changed often.

The same goes for other configs, too.

File: torchtitan/experiments/flux/train.py [hunk 14]
```diff
@@ -155,19 +161,18 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
...
         pred = self._predict_noise(
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: It seems `_predict_noise` doesn't need to be a function in trainer. We can put it as a standalone function.

File: torchtitan/experiments/flux/train.py (hunk 12 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/train.py]: on the line `model_parts = self.model_parts`
you can do
```
assert len(self.model_parts) == 1
model = self.model_parts[0]
```

File: torchtitan/experiments/flux/train.py [hunk 15]
```diff
@@ -204,6 +209,46 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
...
+            or self.step == self.job_config.training.steps
+        ):
+            self._eval_flux_model()
+
+    def _eval_flux_model(self, prompt: str = "A photo of a cat"):
+        """
+        Evaluate the Flux model.
+        1) generate and save images every few steps; 2) Calculate loss with fixed t value on validation set.
+        """
+        self.model_parts[0].eval().requires_grad_(False)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: similar question: why `.requires_grad_(False)` since you already call `eval()`

File: torchtitan/experiments/flux/train.py [hunk 16]
```diff
@@ -204,6 +209,46 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
 
         self.metrics_processor.log(self.step, global_avg_loss, global_max_loss)
 
+        # Evaluate the model during training
+        if (
+            self.step % self.job_config.eval.eval_freq == 0
+            or self.step == self.job_config.training.steps
+        ):
+            self._eval_flux_model()
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: I'd suggest we do
```suggestion
            model.eval()
            self._eval_flux_model()
            model.train()
```
instead of doing it in couple of different places

File: torchtitan/experiments/flux/train.py [hunk 17]
```diff
@@ -204,6 +209,46 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
...
+
+    def _eval_flux_model(self, prompt: str = "A photo of a cat"):
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: maybe rename to `eval_step` to match `train_step`

File: torchtitan/experiments/flux/sampling.py [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+    img_height = 16 * (job_config.eval.sample_img_width // 16)
+    img_width = 16 * (job_config.eval.sample_img_height // 16)
+
+    enable_classifer_free_guidance = job_config.eval.enable_classifer_free_guidance
+
+    # Tokenize the prompt. Unsqueeze to add a batch dimension.
+    clip_tokens = clip_tokenizer.encode(prompt).unsqueeze(0)
+    t5_tokens = t5_tokenizer.encode(prompt).unsqueeze(0)
+
+    batch = preprocess_flux_data(
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/sampling.py]: maybe rename to `preprocess_data`

File: torchtitan/experiments/flux/train.py [hunk 19]
```diff
@@ -204,6 +209,46 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
...
+        """
+        Evaluate the Flux model.
+        1) generate and save images every few steps; 2) Calculate loss with fixed t value on validation set.
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: for 1), let's mention that all DP ranks would run the same prompt, but due to random noise, results could be different and used for comparisons.
for 2), mark it as a TODO item

File: torchtitan/experiments/flux/flux_argparser.py (hunk 10 above)
Comment (wwwjn) [reply torchtitan/experiments/flux/flux_argparser.py]: Sure! Sorry I didn't realize there's another directory path. Let me remove this output_dir here

File: torchtitan/experiments/flux/train_configs/flux_schnell_model.toml (hunk 13 above)
Comment (wwwjn) [reply torchtitan/experiments/flux/train_configs/flux_schnell_model.toml]: The suggested default for flux-schnell model is 4. And for other model (Eg, flux-dev) is 50. The reference number is from here: https://github.com/black-forest-labs/flux/blob/716724eb276d94397be99710a0a54d352664e23b/src/flux/cli.py#L158. I will double check this number as well but I think it's changable for different models.

File: torchtitan/experiments/flux/train.py (hunk 14 above)
Comment (wwwjn) [reply torchtitan/experiments/flux/train.py]: I'm so not sure where to put this function. Should we move the `_predict_noise` as a standalone function in train.py? Or we should put it in utils.py?

File: torchtitan/experiments/flux/train.py (hunk 15 above)
Comment (wwwjn) [reply torchtitan/experiments/flux/train.py]: > similar question: why `.requires_grad_(False)` since you already call `eval()`

Double check the function defintion, don't need to call `requires_grad_` after switching train/eval mode. Removed

File: torchtitan/experiments/flux/train_configs/flux_schnell_model.toml (hunk 13 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/train_configs/flux_schnell_model.toml]: hmm interesting, maybe 50 is a good number for default

File: torchtitan/experiments/flux/train_configs/debug_model.toml [hunk 20]
```diff
@@ -38,15 +38,23 @@ seq_len = 512
 max_norm = 2.0  # grad norm clipping
 steps = 10
 compile = false
-dataset = "cc12m"
-guidance = 3.5
+dataset = "cc12m-wds"
+classifer_free_guidance_prob = 0.1
+img_size = 512
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train_configs/debug_model.toml]:
```suggestion
img_size = 256
```

**Review by tianyu-l — APPROVED** [review]
Looks great!
Please fix linting and img_size before merge.

**Comment by wwwjn** [comment]
> Looks great! Please fix linting and img_size before merge.

Thank you so much for the detailed reply. Fixed.
